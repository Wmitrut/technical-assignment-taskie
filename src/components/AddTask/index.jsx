import { useState } from 'react';
import {
  ActionArea,
  AddButton,
  AddTaskFormWrapper,
  CancelButton,
  DescriptionArea,
  SaveButton,
  TitleInput,
} from './styles';

const AddTask = (props) => {
  const [showNewTaskButton, setShowNewTaskButton] = useState(true);
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');

  const handleInputFinish = () => {
    setShowNewTaskButton(true);
    addNewTask(props.columnId, title, description);
    setTitle('');
    setDescription('');
  };

  const handleCancel = () => {
    setShowNewTaskButton(true);
    setTitle('');
    setDescription('');
  };

  const addNewTask = (columnId, title, description) => {
    const newTaskId = 'task-' + Math.floor(Math.random() * 100000);
    const column = props.board.columns[columnId];
    const newTaskIds = Array.from(column.taskIds);
    newTaskIds.push(newTaskId);

    const newTask = {
      id: newTaskId,
      title: title,
      description: description,
    };

    props.setBoard({
      ...props.board,
      tasks: {
        ...props.board.tasks,
        [newTaskId]: newTask,
      },
      columns: {
        ...props.board.columns,
        [columnId]: {
          ...props.board.columns[columnId],
          taskIds: newTaskIds,
        },
      },
    });
  };

  return (
    <>
      {showNewTaskButton ? (
        <AddButton onClick={() => setShowNewTaskButton(false)} type="button">
          Add new Task
        </AddButton>
      ) : (
        <AddTaskFormWrapper>
          <TitleInput
            type="text"
            placeholder="Title of your new task"
            value={title}
            onChange={(e) => setTitle(e.target.value)}
          />
          <DescriptionArea
            rows={5}
            type="textarea"
            placeholder="describe your new task"
            value={description}
            onChange={(e) => setDescription(e.target.value)}
          />
          <ActionArea>
            <CancelButton onClick={() => handleCancel()} type="button">
              Cancel
            </CancelButton>
            <SaveButton onClick={() => handleInputFinish()} disabled={title === ''} type="button">
              Save
            </SaveButton>
          </ActionArea>
        </AddTaskFormWrapper>
      )}
    </>
  );
};

export default AddTask;
