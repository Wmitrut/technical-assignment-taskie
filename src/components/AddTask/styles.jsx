import styled from 'styled-components';
import { COLORS } from '../../constants/colors';

export const AddButton = styled.button`
  font-familt: Inter;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  width: 100%;
  padding: 15px 5px;
  border: none;
  background: ${COLORS.primary};
  border: 1px solid;
  border-color: ${COLORS.secondary};
  box-sizing: border-box;
  border-radius: 10px;
  color: ${COLORS.light};
  transition: 0.3s all;
  cursor: pointer;

  &:hover {
    background: ${COLORS.primaryDark};
  }
`;

export const AddTaskFormWrapper = styled.div`
  width: 100%;
  padding: 15px;
  background: ${COLORS.primaryDarker};
  box-sizing: border-box;
  border-radius: 5px;
  display: flex;
  flex-direction: column;
`;

export const TitleInput = styled.input`
  font-familt: Inter;
  width: 100%;
  padding: 10px 5px;
  box-sizing: border-box;
  border-radius: 5px;
  border: none;
  margin-bottom: 10px;
  background: ${COLORS.secondary};
  color: ${COLORS.light};
  font-family: Inter;

  &:focus {
    outline: none;
  }
`;

export const DescriptionArea = styled.textarea`
  width: 100%;
  padding: 10px 5px;
  box-sizing: border-box;
  border-radius: 5px;
  border: none;
  margin-bottom: 10px;
  background: #4a4a4a;
  color: ${COLORS.light};
  resize: vertical;
  font-family: Inter;

  &:focus {
    outline: none;
  }
`;

export const ActionArea = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
`;

export const SaveButton = styled.button`
  font-familt: Inter;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  width: 30%;
  padding: 10px 5px;
  border: none;
  background: ${COLORS.primary};
  border: 1px solid;
  border-color: ${COLORS.secondary};
  box-sizing: border-box;
  border-radius: 10px;
  color: ${COLORS.light};
  transition: 0.3s all;
  cursor: pointer;

  &:hover:disabled {
    background: ${COLORS.successDisabled};
    cursor: not-allowed;
  }

  &:hover {
    background: ${COLORS.success};
  }
`;

export const CancelButton = styled.button`
  font-familt: Inter;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  width: 30%;
  padding: 10px 5px;
  border: none;
  background: ${COLORS.primary};
  border: 1px solid;
  border-color: ${COLORS.secondary};
  box-sizing: border-box;
  border-radius: 10px;
  color: ${COLORS.light};
  transition: 0.3s all;
  cursor: pointer;

  &:hover {
    background: ${COLORS.error};
  }
`;
