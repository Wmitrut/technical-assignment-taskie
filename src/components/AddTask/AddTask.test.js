import renderer from 'react-test-renderer';
import AddTask from './index';

it('Renders form correctly', () => {
  const tree = renderer.create(<AddTask />).toJSON();
  expect(tree).toMatchSnapshot();
});
