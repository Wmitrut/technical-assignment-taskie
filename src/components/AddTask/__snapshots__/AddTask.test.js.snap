// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders form correctly 1`] = `
<button
  className="sc-bczRLJ khahOI"
  onClick={[Function]}
  type="button"
>
  Add new Task
</button>
`;
